cpp/src/arrow/compute/exec/tpch_node.cc

```diff
@@ -0,0 +1,3836 @@
+                success &= GenerateVerbPhrase(offset, rng, arr);
+                success &= GenerateNounPhrase(offset, rng, arr);
+                success &= GenerateTerminator(offset, rng, arr);
+                break;
+            case 5:
+                success &= GenerateNounPhrase(offset, rng, arr);
+                success &= GeneratePrepositionalPhrase(offset, rng, arr);
+                success &= GenerateVerbPhrase(offset, rng, arr);
+                success &= GeneratePrepositionalPhrase(offset, rng, arr);
+                success &= GenerateTerminator(offset, rng, arr);
+                break;
+            default:
+                Unreachable("Random number should be between 0 and 5 inclusive");
+                break;
+            }
+            return success;
+        }
+
+        using GenerateColumnFn = std::function<Status(size_t)>;
+        class PartAndPartSupplierGenerator
+        {
+        public:
+            Status Init(
+                size_t num_threads,
+                int64_t batch_size,
+                float scale_factor)
+            {
+                if(!inited_)
+                {
+                    inited_ = true;
+                    batch_size_ = batch_size;
+                    scale_factor_ = scale_factor;
+
+                    thread_local_data_.resize(num_threads);
+                    for(ThreadLocalData &tld : thread_local_data_)
+                    {
+                        // 5 is the maximum number of different strings we need to concatenate
+                        tld.string_indices.resize(5 * batch_size_);
+                    }
+                    part_rows_to_generate_ = static_cast<int64_t>(scale_factor_ * 200000);
+                }
+                return Status::OK();
+            }
+            
+            int64_t part_batches_generated() const
+            {
+                return part_batches_generated_.load();
+            }
+
+            int64_t partsupp_batches_generated() const
+            {
+                return partsupp_batches_generated_.load();
+            }
+
+            Result<std::shared_ptr<Schema>> SetPartOutputColumns(const std::vector<std::string> &cols)
+            {
+                return SetOutputColumns(cols, part_types_, part_name_map_, part_cols_);
+            }
+
+            Result<std::shared_ptr<Schema>> SetPartSuppOutputColumns(const std::vector<std::string> &cols)
+            {
+                return SetOutputColumns(cols, partsupp_types_, partsupp_name_map_, partsupp_cols_);
+            }
+
+            Result<util::optional<ExecBatch>> NextPartBatch()
+            {
+                size_t thread_index = thread_indexer_();
+                ThreadLocalData &tld = thread_local_data_[thread_index];
+                {
+                    std::lock_guard<std::mutex> lock(part_output_queue_mutex_);
+                    bool all_generated = part_rows_generated_ == part_rows_to_generate_;
```

Comment:
You can't move that if above because we may have generated all rows, but the queue may not be empty, so we need to check the queue first, and only then return if we've generated everything. I did get rid of the variable though and just pulled it into the `else if`.